### PR TITLE
Factor waker out of `dreadnought` and add `task_join`

### PR DESCRIPTION
**Pull request by tsoutsman**
The waker model can be used to implement blocking calls (e.g. `join`), and so it shouldn't be defined in `dreadnought`.

I hit a circular dependency problem similar to #766. `task_join::join`, relies on `waker`, which in turn relies on `scheduler` and `mutex_sleep`, both of which depend on `task`.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

### CODE REVIEW

File: kernel/waker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+}
+
+/// A waker that unblocks the given task when awoken.
+#[derive(Debug)]
+struct Waker {
+    /// Whether the waker has been activated.
+    ///
+    /// This field ensures `execute` detects if the waker was activated prior to
```
Comment (NathanRoyer) [inline kernel/waker/src/lib.rs]: What is this `execute` mentioned in this comment?

Comment (tsoutsman) [reply kernel/waker/src/lib.rs]: Nice catch. Meant to say `Blocker::block`.

File: kernel/waker/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            activated: activated.clone(),
+            task: get_my_current_task().expect("failed to get current task"),
+        })),
+        Blocker { inner: activated },
+    )
+}
+
+/// A blocker that blocks the current task until the associated waker is woken.
+pub struct Blocker {
+    inner: Arc<Mutex<bool>>,
```
Comment (kevinaboos) [inline kernel/waker/src/lib.rs]: should probably add a comment here explaining why we use a `Mutex<bool>` instead of an `AtomicBool`

**Review by kevinaboos — COMMENTED** [review]
I like this, but it is a bit messy and likely confusing to newcomers why there needs to be a separate `task_join` crate.

Ideally, the action of blocking the current task while it is waiting to `join` another task would take place within `JoinableTaskRef::join()` itself. That would make the action of setting the current task's `waker` internal to the call to `join()`, which makes sense IMO from an API perspective. Is there any other use case where you would want to `set_waker` for a task but _not_ also call `join`?


I wonder how monolithic the `task` crate would have to become if we wanted to achieve this; there is likely some tension between that goal and our other goal of keeping crates small and plentiful.

**Comment by kevinaboos** [comment]
Once #879 is merged in, we can use that here for simplicity.

**Review by kevinaboos — APPROVED** [review]

**Pull request merged**